xma_data_processor.py

```diff
@@ -0,0 +1,374 @@
+"""Converts data from XMALab into the format useful for DLC training."""
+
+import os
+
+import blend_modes
+import cv2
+import numpy as np
+import pandas as pd
+from ruamel.yaml import YAML
+
+
+class XMADataProcessor:
+    """Converts data from XMALab into the format useful for DLC training."""
+
+    def __init__(self, config):
+        self._config = config
+        self._mode = config["tracking_mode"]
+
+    def get_bodyparts_from_xma(
+        self, path_to_trial, split_markers=False, crossed_markers=False
+    ):
+        """Pulls the names of the XMAlab markers from the 2Dpoints file"""
+
+        csv_path = [file for file in os.listdir(path_to_trial) if file[-4:] == ".csv"]
+        if len(csv_path) > 1:
+            raise FileExistsError(
+                "Found more than 1 CSV file for trial: " + path_to_trial
+            )
+        if len(csv_path) <= 0:
+            raise FileNotFoundError(
+                "Couldn't find a CSV file for trial: " + path_to_trial
+            )
+        trial_csv = pd.read_csv(
+            path_to_trial + "/" + csv_path[0],
+            sep=",",
+            header=0,
+            dtype="float",
+            na_values="NaN",
+        )
+        names = trial_csv.columns.values
+        if self._mode == "rgb":
+            parts = [name.rsplit("_", 1)[0] for name in names]
+            if split_markers:
+                parts = parts + [f"sw_{part}" for part in parts]
+            if crossed_markers:
+                parts = parts + [
+                    f"cx_{part}_cam1x2"
+                    for part in [name.rsplit("_", 2)[0] for name in names]
+                ]
+        elif self._mode in ["2D", "per_cam"]:
+            parts = [name.rsplit("_", 2)[0] for name in names]
+        else:
+            raise SyntaxError("Invalid value for mode parameter")
+
+        parts_unique = []
+        for part in parts:
+            if part not in parts_unique:
+                parts_unique.append(part)
+        return parts_unique
+
+    def make_rgb_video(self, data_path):
+        """For all trials in given data path merges 2 videos into single RBG video."""
+        trials = [
+            folder
+            for folder in os.listdir(data_path)
+            if os.path.isdir(os.path.join(data_path, folder))
+            and not folder.startswith(".")
+        ]
+        for trial in trials:
+            path_to_trial = f"{data_path}/{trial}"
+            self._merge_rgb(path_to_trial)
+            substitute_data_relpath = "labeled-data/" + self._config["dataset_name"]
+            substitute_data_abspath = os.path.join(
+                os.path.split(self._config["path_config_file"])[0],
+                substitute_data_relpath,
+            )
+            self._extract_matched_frames_rgb(
```

Comment:
You may also want to decouple the behavior here for videos vs. data points/image extraction. extract_matched_frames works for trials with an XMAlab-formatted CSV attached (typically training data), while merge_rgb works on all videos as long as there's a cam 1 and a cam2 video.

Reply:
added a TODO, to keep changes smaller / more focused